### DISCUSSION TITLE
Object Manipulation Functions

### DISCUSSION THREAD

**Discussion by innovate-invent** [jmespath-community/jmespath.spec#47] [Q&A] [answered]
We need a mechanism for building objects with an unknown number of keys. This has already been extensively discussed and this discussion is to summarise, preparing to create a JEP.

https://github.com/jmespath/jmespath.py/pull/105
https://github.com/jmespath/jmespath.site/issues/17

The proposed functions are:
```
array[array[string, any]] items(object)
array[array[any, any]] zip([array *argument, [, array $...]])
object from_items(array[array[string, any]])
```

`items()` takes an object and returns an array of pairs [key, value] for each entry
`zip()` behaves exactly like the python zip() function and allows combining an array of keys and array of values into what is required by the next function
`from_items()` takes an array of pairs [key, value] and builds an object with the respective key->value associations

`zip()` is a useful operation for other uses and can accept more than two arrays to combine.
The arrays returned by the existing `keys()` and `values()` functions can be combined using `zip()` to reproduce the original object via `from_items()`

examples:
`search({"a":1, "b":2}, items(@))` -> `[["a", 1], ["b", 2]]`
`search([["a", 1], ["b", 2]], from_items(@))` -> `{"a":1, "b":2}`
`search({"a":[1, 2, 3], "b":[4,5,6], "c":[7,8,9]}, zip(a, b, c))` -> `[[1,4,7], [2,5,8], [3,6,9]]`

an alternative to `items()` could be `search({"a":1, "b":2}, let({"k":keys(@), "v":values(@)}, &map(&[k[@], v[@]], range(length(@)))))` where `range()` is the same as the python function by the same name.

an alternative to `zip()` could be `search({"a":[1, 2, 3], "b":[4,5,6], "c":[7,8,9]}, let({a, b, c}, &map(&[a[@], b[@], c[@]], range(len(a)))))`
This pattern allows specifying which array determines the final length, while the index accessors return null for any list shorter than the specified one.

an alternative to `from_items()` could be `search([["a", 1], ["b", 2]], merge(map(&{@[0] : @[1]}, @)))`
This requires expanding the grammar to allow expressions in place of the key. `merge()` also needs to allow accepting a single array argument and operating over that as if it were its argument list.

I think with these alternatives, we need to discuss how strictly we want to adhere to the "one way of doing something" guideline. Where do we draw the line between a wiki of reusable patterns and adding functions to simplify?

**Comment by innovate-invent** [comment]
The more I think about this, the more I am worried we are adding functions that should actually be syntax.

This is most apparent when considering `concat()` vs `merge()`, why does `concat()` exist as syntax and `merge()` doesn't?
`concat()` is implemented as the flatten operation: `[[1], [2]][]` -> `[1,2]`
should this be extended to include objects?: `[{"a":1}, {"a":2, "b":1}][]` -> `{"a":2, "b":1}`
we could add syntax such that it would be: `[{"a":1}, {"a":2, "b":1}]{}` -> `{"a":2, "b":1}`

alternatively we could look at adding a spread operator to multiselect list and hash:
`[[1].., [2]..]` -> `[1,2]`
`{{"a":1}.., {"a":2, "b":1}..}` -> `{"a":2, "b":1}`

We also have two different ways of applying the `map()` operation: `map(&to_string(@), k)` vs `k[*].to_string(@)`

If we go this route then the alternatives would be:
items: `let({"k":keys(@), "v":values(@)}, range(length(@))[*].[k[@], v[@]])` or `let({a: @}, keys(@)[*].[@, a[@]])`
zip: `let({a, b, c}, range(len(a))[*].[a[@], b[@], c[@]])`
from_items: `[*].{[0] : [1]} | []`

or with the [reduce operation](https://github.com/jmespath-community/jmespath.jep/discussions/48)
from_items: `[&].[acc, {cur[0]:cur[1]}]{}`

or with the spread operator:
from_items: `[&].{acc.., cur[0]: cur[1]}`

or with the [set union operator](https://github.com/jmespath-community/jmespath.jep/discussions/25#discussioncomment-2378378):
from_items: `[&].(acc + {cur[0]: cur[1]})`

**Reply by innovate-invent** [reply]
The from_items examples would require modifying the `keyval-expr = identifier ":" expression` syntax to be `keyval-expr = expression ":" expression`, or possibly `keyval-expr = (paren-expression / identifier) ":" expression` for backwards compatibility. Allowing dynamic keys would be useful in other situations.

**Comment by springcomp** [comment, marked as answer]
# Object Manipulation Functions

|||
|---|---
| **JEP**    |  13
| **Author** | Jonathan Stewmon
| **Created**| 21-March-2016
| **SemVer** | MINOR
| **Status**| Draft
| **[Discussion #47]** | #47

## Abstract

As a JSON querying language, JMESPath has long needed some functions to manipulate JSON objects.

This JEP introduces built-in functions to extract a list of key/value pairs from a JSON object and conversely converting an array of key/value pairs to a JSON object.

## Specification

### items

```
array[array[any]] items(object $obj)
```

Returns a an array of key value pairs for the provided object `$obj`. Each pair is a 2-item array with the first item being the key and the second item being the value. This function is the inverse of the `from_items()` function.

Note that because JSON hashes are inheritently unordered, the key value pairs of the provided object `$obj` are inheritently unordered.  Implementations are not required to return values in any specific order. 

For example, given the input:

```json
{"a": "first", "b": "second", "c": "third"}
```

The expression ``items(@)`` could have any of these return values:

- ``[["a", "first"], ["b", "second"], ["c", "third"]]``
- ``[["a", "first"], ["c", "third"], ["b", "second"]]``
- ``[["b", "second"], ["a", "first"], ["c", "third"]]``
- ``[["b", "second"], ["c", "third"], ["a", "first"]]``
- ``[["c", "third"], ["a", "first"], ["b", "second"]]``
- ``[["c", "third"], ["b", "second"], ["a", "first"]]``

If you would like a specific order, consider using the ``sort_by`` function.

### Examples

|Given|Expression|Result
|---|---|---
|``{"a": "first", "b": "second"}``|``items(@)``|``[["b", "second"], ["a", "first"]]``
|``{"z": "last", "b": "second"}``|``sort_by(items(@), &[0])``|``[["b", "second"], ["z", "last"]]``
|``{"z": "last", "b": "second"}``|``sort_by(items(@), &[1])``|``[["z", "last"], ["b", "second"]]``

### from_items

```
object from_items(array[array[any]] $arg)
```
Returns an object from the provided array of key value pairs. This function is the inverse of the `items()`.

### Examples

|Given|Expression|Result
|---|---|---
|``[["one", 1], ["two", 2]]``|``from_items(@)``|``{"one": 1, "two": 2}``
|``[["one", 1], ["two", 2], ["one", 3]]``|``from_items(@)``|``{"one": 3, "two": 2}``

### zip

```
array[array[any]] zip([array[any] $arg, [, array[any] $...]])
```

Accepts one or more arrays as arguments and returns an array of arrays in which the *i-th* array contains the *i-th* element from each of the argument arrays. The returned array is truncated to the length of the shortest argument array.

### Examples

|Expression|Result
|---|---
|``zip(`["a", "b"]`, `[1, 2]`)``|``[["a", 1], ["b", 2]]``
|``zip(`["a", "b", "c"]`, `[1, 2]`)``|``[["a", 1], ["b", 2]]``

## Compliance tests

A new `objects.json` file will be added to the compliance test suite.

## History

N/A

**Reply by gibson042** [reply]
> Implementations are not required to return values in any specific order.

JavaScript learned the hard way that this is not really a tenable position for a language with multiple implementations, because [in practice](https://www.hyrumslaw.com/) a popular one with a predictable ordering becomes the de facto standard against which authors base their expectations and report bugs to other ones that don't match it. Instead, this should be updated to either specify a deterministic order or to _require_ unpredictability.

**Reply by springcomp** [reply]
What about rewording this to:

`The order in which values are returned is therefore undefined.`

**Reply by gibson042** [reply]
That suffers from exactly the same problem. The order _should_ be defined, as either explicitly specified or as intentionally unpredictable (such that an implementation which sorted entries or preserved source order would be nonconforming).

**Reply by springcomp** [reply]
@gibson042  Interesting take.

I’m not sure I agree, though.

If the oder is _undefined_, we do not impose _any_ constraint on any implementation.
Which means that an implementation that chooses to sort them would still be compliant.
As would be an implementation that chooses to preserve source order.

In software term, _undefined_ [already means unpredictable](https://en.wikipedia.org/wiki/Undefined_behavior).

I think the gist of the text (of which I’m not the original author) is to warn consumers to not rely on any order. So it sort of touches on the unpredictability of the result. In practice, the unpredictability is a feature that renders the code portable between implementations.

I would be happy to add text to that effect.

If there is strong feedback advising to explicitely impose a specific order in the spec, I would be happy to add that too of course.

However, I feel that imposing strict unpredictability is a step too far because it would mandate implementations to apply a random sort to their results. Also, it would incidentally make writing the compliance test suite effectively impossible.

**Reply by innovate-invent** [reply]
I think it would be erroneous for an implementation to sort or randomise a mapping. Document/insertion order is an important ordering that should be preserved.

Golang does not guarantee order and it is a hair pulling experience trying to deserialise and then reserialise a document in a go template.
Golang does provide an OrderedMapping construct, but if a library chooses to not use it I then have to start patching around it.

**Reply by gibson042** [reply]
@springcomp That reasoning is not strictly wrong, but the problem is that it breaks down if a sufficiently popular implementation _does_ provide results in a predictable order, because what happens then is that users come to rely upon that behavior (cf. [Hyrum's Law](https://www.hyrumslaw.com/)) and less popular implementations are forced to follow suit, resulting in a de facto predictable order that eventually gets added to the specification anyway (which is exactly what happened in JavaScript). I am only aware of two ways around this: either specify a predictable order up front, or specify that order is intended to be unpredictable in any particular implementation (which is indeed basically impossible to ensure with conformance tests).

I absolutely understand the desire to let the use of predictable vs. unpredictable ordering be unspecified and therefore support maximum convenience across implementations, but my experience is that it just doesn't work out in the long run.

**Reply by springcomp** [reply]
@jstewmon I took the liberty to use [your pull request](https://github.com/jmespath/jmespath.py/pull/105) and turn this into a [JEP](https://github.com/jmespath-community/jmespath.spec/discussions/47#discussioncomment-3308897). You may want to share your feedback when discussing this JEP.

@gibson042 words to the effect that _implementations are required to preserve source order_ would not bother me so much. Would that be OK for you?

**Reply by jstewmon** [reply]
👋 Hi, thanks for thinking of me. I'm not currently using jmespath in a project, so I don't feel inclined to argue any points strongly, but I'll leave a few remarks on how I see things. 😄 

I think everything in the parent post looks good.

IMO, the ordering behavior of `items` should be _consistent_ with `keys` and `values`:
https://github.com/jmespath-community/jmespath.spec/blob/4a8c3234d32af69b10f239cfc567255138fcdfab/functions/keys.yml#L13-L17
https://github.com/jmespath-community/jmespath.spec/blob/4a8c3234d32af69b10f239cfc567255138fcdfab/functions/values.yml#L13-L44

I think the originally proposed description (below) is consistent with the description of `keys`:
> Note that because JSON hashes are inheritently unordered, the key value pairs of the provided object $obj are inheritently unordered. Implementations are not required to return values in any specific order.

I don't think it's appropriate to debate the ordering semantics of `items` in this proposal. If there's cause to debate the ordering semantics of _all_ the functions that return an array from an object, make a separate proposal and debate it there.

**Reply by gibson042** [reply]
> IMO, the ordering behavior of `items` should be _consistent_ with `keys` and `values`:
> 
> I don't think it's appropriate to debate the ordering semantics of `items` in this proposal. If there's cause to debate the ordering semantics of _all_ the functions that return an array from an object, make a separate proposal and debate it there.

Excellent point. Spun off to https://github.com/jmespath-community/jmespath.spec/discussions/94 .

**Reply by springcomp** [reply]
Good. Thanks a lot for chiming in @jstewmon . Your feedback was precious.

**Reply by springcomp** [reply]
For information, those functions [are demonstrated here](https://jmespath.site/#wiki-object-manipulation-functions).

**Comment by springcomp** [comment]
Should `zip()` be included in this proposal?

**Reply by innovate-invent** [reply]
~~That depends on if we allow dynamic key values. `reduce()` would be able to accomplish it with minimal syntax.~~

**Reply by springcomp** [reply]
@innovate-invent I’m not sure what discussion you are referring to. Can you elaborate?

**Reply by innovate-invent** [reply]
haha, I was literally typing a correction as you posted this.

it isn't the reduce operation that competes, it is the [arithmetic-operators](https://github.com/jmespath-community/jmespath.spec/discussions/25#discussioncomment-2378378).

zip() is equivalent to the cartesian product `[a,b] * [1,2]` -> `[[a,1],[b,2]]` or even `[a,b] * [1,2] * [c,d]` -> `[[a,1,c],[b,2,d]]`

**Reply by springcomp** [reply]
Ah I see. Maybe we need to settle on basic arithmetic operations first, before moving on to vectors and matrices 😊
I think `zip()` does not necessarily belongs to this JEP.

**Reply by jstewmon** [reply]
> zip() is equivalent to the cartesian product `[a,b] * [1,2]` -> `[[a,1],[b,2]]` or even `[a,b] * [1,2] * [c,d]` -> `[[a,1,c],[b,2,d]]`

This is not true.

The [cartesian product] has a length equal to the product of the input lengths. The cartesian product of `[a, b] * [1, 2] = [[a, 1], [a, 2], [b, 1], [b, 2]]`.

The `zip` result has length equal to the min length of its inputs. The `zip` function can be though of as correlating the items from each input list by their ordinal position. `zip([a, b], [1, 2]) = [[a, 1], [b, 2]]`.

`zip` is commonly used to produce an argument for `from_items`. Consider tabular data, like a database query result set or CSV - the field labels are commonly defined once  (e.g. the first row, or a separate data structure) in a sequence with ordinal positions corresponding to the position of their values in the data rows. If you want to produce an object/map from each data row, then you can call `from_items(zip(fields, values))` where `values` is a data row.

[cartesian product]: https://en.wikipedia.org/wiki/Cartesian_product

**Reply by innovate-invent** [reply]
sorry, you are right!
I was struggling to find my example of an alternative to zip() that I had previously posted, saw my proposal of the product of objects and just went with it without looking too close.

The actual competing syntax would be `search({x: [a,b], y:[1,2]}, let({y: y}}, x[*].[@, y[$]])` which is using let(), [dynamic subscription](https://github.com/jmespath-community/jmespath.spec/discussions/74), and [`$`](https://github.com/jmespath-community/jmespath.spec/discussions/18)

I do want to explore a operator syntax alternative to zip though

**Reply by jstewmon** [reply]
Personally, I find that dynamic subscription syntax inscrutable. 😅  Zip may not be ubiquitous, but it is familiar from the Python and Ruby standard libraries, lodash for JavaScript, etc.

Anyway, like I said in my other comment, I'm not really using jmespath anymore, so I don't have a stake in the outcome.
